File: client/src/main/java/PuzzleBoard.java
import javax.imageio.ImageIO;
import javax.swing.*;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.awt.image.CropImageFilter;
import java.awt.image.FilteredImageSource;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.net.URLConnection;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.IntStream;

@SuppressWarnings("serial")
public class PuzzleBoard extends JFrame {
	
	final int rows, columns;
	private List<Tile> tiles = new ArrayList<>();
	
	private SelectionManager selectionManager = new SelectionManager();
	
    public PuzzleBoard(final int rows,
                       final int columns,
                       final String imagePath,
                       final String playerid,
                       final String puzzleid) throws IOException {
    	this.rows = rows;
		this.columns = columns;
    	
    	setTitle("Puzzle");
        setResizable(false);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        
        final JPanel board = new JPanel();
        board.setBorder(BorderFactory.createLineBorder(Color.gray));
        board.setLayout(new GridLayout(rows, columns, 0, 0));
        getContentPane().add(board, BorderLayout.CENTER);
        final JPanel info = new JPanel();
        info.setLayout(new GridLayout(2,2));
        final JLabel playerIdLabel = new JLabel("PlayerID:");
        final JLabel playerId = new JLabel(playerid);
        final JLabel puzzleIdLabel = new JLabel("PuzzleID:");
        final JLabel puzzleId = new JLabel(puzzleid);
        info.add(playerIdLabel);
        info.add(playerId);
        info.add(puzzleIdLabel);
        info.add(puzzleId);
        getContentPane().add(info, BorderLayout.SOUTH);
        
        createTiles(imagePath);
        paintPuzzle(board);
    }

    
    private void createTiles(final String imagePath) throws IOException {

        URL url = new URL(imagePath);

        URLConnection conn = url.openConnection();
        conn.setRequestProperty("User-Agent", "Mozilla/5.0");

        conn.connect();
        InputStream urlStream = conn.getInputStream();
        final BufferedImage image ;

        try {
            image = ImageIO.read(url);
        } catch (IOException ex) {
            JOptionPane.showMessageDialog(this, "Could not load image", "Error", JOptionPane.ERROR_MESSAGE);
            return;
        }

        final int imageWidth = image.getWidth(null);
        final int imageHeight = image.getHeight(null);

        int position = 0;
        
        final List<Integer> randomPositions = new ArrayList<>();
        IntStream.range(0, rows*columns).forEach(item -> { randomPositions.add(item); });
        Collections.shuffle(randomPositions);
        
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < columns; j++) {
            	final Image imagePortion = createImage(new FilteredImageSource(image.getSource(),
                        new CropImageFilter(j * imageWidth / columns,
                        					i * imageHeight / rows, 
                        					(imageWidth / columns), 
                        					imageHeight / rows)));

                tiles.add(new Tile(imagePortion, position, randomPositions.get(position)));
                position++;
            }
        }
	}
    
    private void paintPuzzle(final JPanel board) {
    	board.removeAll();
    	
    	Collections.sort(tiles);
    	
    	tiles.forEach(tile -> {
    		final TileButton btn = new TileButton(tile);
            board.add(btn);
            btn.setBorder(BorderFactory.createLineBorder(Color.gray));
            btn.addActionListener(actionListener -> {
            	selectionManager.selectTile(tile, () -> {
            		paintPuzzle(board);
                	checkSolution();
            	});
            });
    	});
    	
    	pack();
        setLocationRelativeTo(null);
    }

    private void checkSolution() {
    	if(tiles.stream().allMatch(Tile::isInRightPlace)) {
    		JOptionPane.showMessageDialog(this, "Puzzle Completed!", "", JOptionPane.INFORMATION_MESSAGE);
    	}
    }
}

File: client/src/main/kotlin/it/unibo/pcd/client/ui/PuzzleBoard.kt
package it.unibo.pcd.client.ui


import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import it.unibo.pcd.client.controller.ViewController
import java.awt.BorderLayout
import java.awt.Color
import java.awt.GridLayout
import java.awt.event.ActionEvent
import java.awt.image.BufferedImage
import java.awt.image.CropImageFilter
import java.awt.image.FilteredImageSource
import java.io.IOException
import java.net.URL
import java.util.*
import java.util.function.Consumer
import java.util.function.IntFunction
import java.util.stream.Collectors
import java.util.stream.IntStream
import javax.imageio.ImageIO
import javax.swing.*


class PuzzleBoard(private val rows: Int,
                  private val columns: Int,
                  private val imagePath: String,
                  private val playerid: String,
                  private val puzzleid: String,
                  private val state: JsonArray,
                  private val controller: ViewController) : JFrame() {

    private var tiles: MutableList<Tile> = ArrayList()
    private val info = JPanel()
    private val playerIdLabel = JLabel()
    private val puzzleIdLabel = JLabel()
    private val playerId = JLabel()
    private val puzzleId = JLabel()
    private val selectionManager: SelectionManager = SelectionManager(controller)
    val board = JPanel()
    private var currentState = mutableListOf<Int>()

    init {
        title = "Puzzle"
        isResizable = false
        defaultCloseOperation = EXIT_ON_CLOSE

        contentPane.add(board, BorderLayout.CENTER)
        info.layout = GridLayout(2, 2)
        playerIdLabel.text = "PlayerID:"
        playerId.text = playerid
        puzzleIdLabel.text = "PuzzleID:"
        puzzleId.text = puzzleid
        info.add(playerIdLabel)
        info.add(playerId)
        info.add(puzzleIdLabel)
        info.add(puzzleId)

        contentPane.add(info, BorderLayout.SOUTH)
        createTiles(imagePath)
        board.border = BorderFactory.createLineBorder(Color.gray)
        board.layout = GridLayout(rows, columns, 0, 0)
        paintPuzzle(board)
    }

    @Throws(IOException::class)
    private fun createTiles(imagePath: String) {
        val url = URL(imagePath)
        val conn = url.openConnection()
        conn.setRequestProperty("User-Agent", "Mozilla/5.0")
        conn.connect()
        val urlStream = conn.getInputStream()
        val image: BufferedImage
        image = try {
            ImageIO.read(url)
        } catch (ex: IOException) {
            JOptionPane.showMessageDialog(this,
                    "Could not load image", "Error", JOptionPane.ERROR_MESSAGE)
            return
        }
        val imageWidth = image.getWidth(null)
        val imageHeight = image.getHeight(null)
        var position = 0
        val randomPositions = IntStream.range(0, state.size())
                .mapToObj(IntFunction<Any> { i: Int ->
                    state.getInteger(i)
                }).collect(Collectors.toList<Any>())
        for (i in 0 until rows) {
            for (j in 0 until columns) {
                val imagePortion = createImage(FilteredImageSource(image.source,
                        CropImageFilter(j * imageWidth / columns,
                                i * imageHeight / rows,
                                imageWidth / columns,
                                imageHeight / rows)))

                tiles.add(Tile(imagePortion, position, randomPositions[position] as Int))
                position++
            }
        }

        val copy: MutableList<Tile> = ArrayList()
        for (i in 0 until tiles.size) {
            copy.add(i, tiles[state.getInteger(i)])
        }
        tiles = copy
        for (i in 0 until state.size()) {
            currentState.add(i, state.getInteger(i))
        }

        for (i in 0 until tiles.size) {
            tiles[i].currentPosition = currentState[i]
        }
    }

    fun paintPuzzle(board: JPanel) {
        var selectionActive = false
        var selectedTile: Tile? = null

        board.removeAll()
        tiles.forEach(Consumer { tile: Tile? ->
            val btn = TileButton(tile!!)
            board.add(btn)
            btn.border = BorderFactory.createLineBorder(Color.gray)
            btn.addActionListener {
                if (selectionActive) {
                    selectionActive = false
                    val message = JsonObject()
                            .put("playerid", playerId.text)
                            .put("puzzleid", puzzleId.text)
                            .put("source", selectedTile!!.currentPosition.toString())
                            .put("destination", tile.currentPosition.toString())

                    swapTile(selectedTile, tile)
                    paintPuzzle(board)
                    controller.swap(message)
                } else {
                    selectionActive = true
                    selectedTile = tile
                }
            }
        })
        pack()
        setLocationRelativeTo(null)
    }

    fun swapTile(t1: Tile?, t2: Tile) {

        val pos = t1!!.currentPosition
        t1.currentPosition = (t2.currentPosition)
        t2.currentPosition = pos
        val first = t1.currentPosition
        val second = t2.currentPosition
        val tmp = first
        currentState[currentState.indexOf(first)] = second
        currentState[currentState.indexOf(second)] = tmp
        tiles.sort()
    }


    fun repaintPuzzle(newState: JsonObject) {

        SwingUtilities.invokeLater {
            val state = newState.getJsonArray("state")
            val newState = mutableListOf<Int>()
            for (i in 0 until state.size()) {
                newState.add(state.getInteger(i))
            }

            val difference = mutableListOf<Int>()
            for (i in 0 until currentState.size) {
                if (currentState[i] != newState[i]) {
                    difference.add(currentState[i])
                }
            }
            if (difference.isNotEmpty() || !difference.isNullOrEmpty()) {
                currentState = newState
                selectionManager.swapTile(tiles[difference[0]], tiles[difference[1]])
                paintPuzzle(board)
            }
        }
    }
}